File: packages/flint-js/src/vegalite/templates/violin.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

/**
 * Vega-Lite Violin Plot template.
 *
 * A violin plot shows the **full smoothed distribution shape** of a quantitative
 * measure, one mirrored kernel-density curve per category. It is the
 * density-plot cousin of the Boxplot: where the boxplot draws quartiles, the
 * violin draws the whole KDE, symmetric about a center line so the *width*
 * encodes how many observations fall near each value. The read is the density
 * area/width (bimodality, skew, spread), not a position — hence the 'area'
 * cognitive channel (mirrors density.ts).
 *
 * Contract (mirrors Boxplot's channel mapping):
 *   x      — the category (discrete grouping). One violin per x value.
 *   y      — the quantitative measure whose distribution is drawn (the shared
 *            continuous "value" axis).
 *   color  — optional; defaults to the category so each violin gets its own hue.
 *   row    — optional OUTER facet (small multiples of the whole violin panel).
 *
 * Vega-Lite native idiom (no plugins): VL's `density` transform per category
 * plus a mirrored area (`x = density, stack: "center"`). The canonical VL violin
 * places each category in its own **column facet** with the measure on `y` and
 * the mirrored density on `x`. This template therefore CONSUMES the column-facet
 * slot for the per-category panels (the category supplied on `x` is moved into a
 * VL column/wrap facet). `column` is consequently NOT a user-available channel —
 * an additional outer facet is offered through `row` only. The measure stays on
 * the shared `y` (value) axis so all violins are directly comparable.
 *
 * Reuses density.ts's bandwidth wiring and the boxplot's discrete-axis handling.
 */

import { ChartTemplateDef, ChartPropertyDef } from '../../core/types';
import { detectBandedAxisForceDiscrete } from './utils';

const isDiscrete = (t: string | undefined) => t === 'nominal' || t === 'ordinal';

/** Distinct non-null values of a field, in data-encounter order. */
function distinctValues(table: any[], field: string): any[] {
    const seen = new Set<string>();
    const out: any[] = [];
    for (const row of table) {
        const v = row[field];
        if (v == null) continue;
        const key = String(v);
        if (!seen.has(key)) { seen.add(key); out.push(v); }
    }
    return out;
}

/** [min, max] of a numeric field across the table (ignoring non-numbers). */
function numericExtent(table: any[], field: string): [number, number] | null {
    let min = Infinity, max = -Infinity;
    for (const row of table) {
        const v = row[field];
        if (typeof v !== 'number' || !isFinite(v)) continue;
        if (v < min) min = v;
        if (v > max) max = v;
    }
    if (min === Infinity || max === -Infinity) return null;
    if (min === max) { min -= 0.5; max += 0.5; }
    return [min, max];
}

/** Linear-interpolated quantile of an already-sorted ascending array. */
function quantileSorted(sorted: number[], p: number): number {
    const n = sorted.length;
    if (n === 0) return NaN;
    const idx = (n - 1) * p;
    const lo = Math.floor(idx), hi = Math.ceil(idx);
    if (lo === hi) return sorted[lo];
    return sorted[lo] + (sorted[hi] - sorted[lo]) * (idx - lo);
}

/** Sample standard deviation (n-1) of a numeric array. */
function stdev(values: number[]): number {
    const n = values.length;
    if (n < 2) return 0;
    const mean = values.reduce((s, x) => s + x, 0) / n;
    const v = values.reduce((s, x) => s + (x - mean) * (x - mean), 0) / (n - 1);
    return Math.sqrt(v);
}

/**
 * Normal-reference (Scott/Silverman) KDE bandwidth — the same rule Vega's
 * `density` transform uses when no bandwidth is given. Lets us pad the shared
 * density `extent` by a multiple of the *actual* kernel width so every violin
 * tapers to ~zero at its ends instead of being clipped flat (see instantiate).
 */
function bandwidthNRD(values: number[]): number {
    const n = values.length;
    if (n < 2) return 0;
    const s = [...values].sort((a, b) => a - b);
    const lo = quantileSorted(s, 0.25);
    const hi = quantileSorted(s, 0.75);
    const sd = stdev(s);
    let h = Math.min(sd, (hi - lo) / 1.34);
    if (!(h > 0)) h = sd || Math.abs(lo) || 1;
    return 1.06 * h * Math.pow(n, -0.2);
}

/**
 * Largest per-group KDE bandwidth across the groups Vega will form (one per
 * distinct `groupby` tuple). Vega computes the auto bandwidth per group, so the
 * widest-spread group dictates how far the shared extent must reach to taper.
 */
function maxGroupBandwidth(table: any[], measure: string, groupby: string[]): number {
    const groups = new Map<string, number[]>();
    for (const row of table) {
        const v = row[measure];
        if (typeof v !== 'number' || !isFinite(v)) continue;
        const key = groupby.map((f) => String(row[f])).join('\u0000');
        let arr = groups.get(key);
        if (!arr) { arr = []; groups.set(key, arr); }
        arr.push(v);
    }
    let max = 0;
    for (const arr of groups.values()) {
        const bw = bandwidthNRD(arr);
        if (bw > max) max = bw;
    }
    return max;
}

export const violinPlotDef: ChartTemplateDef = {
    chart: 'Violin Plot',
    template: {
        mark: { type: 'area', orient: 'horizontal' },
        transform: [{ density: '__measure__', groupby: [], as: ['value', 'density'] }],
        encoding: {
            // Measure → shared continuous value axis (vertical).
            y: { field: 'value', type: 'quantitative' },
            // Mirrored kernel density → horizontal width, centered (the violin).
            x: {
                field: 'density', type: 'quantitative', stack: 'center',
                impute: null, title: null,
                axis: { labels: false, ticks: false, grid: false },
            },
        },
    },
    // `column` is consumed internally for the per-category panels; only `row`
    // is exposed as an additional outer facet.
    channels: ['x', 'y', 'color', 'row'],
    markCognitiveChannel: 'area',
    declareLayoutMode: (cs, table) => {
        // The category lives on `x`; force it discrete (boxplot-style) so a
        // numeric/temporal category still resolves to clean bands/panels.
        if (!cs.x?.field || !cs.y?.field) return {};
        const result = detectBandedAxisForceDiscrete(cs, table, { preferAxis: 'x' });
        if (!result) return {};
        return { resolvedTypes: result.resolvedTypes };
    },
    instantiate: (spec, ctx) => {
        const { x, y, color, row } = ctx.resolvedEncodings;
        const catField = x?.field;
        const measureField = y?.field;
        if (!catField || !measureField) return;

        // --- Density transform (per category, retaining every facet field) ---
        // The density transform only keeps its `groupby` fields on the output
        // rows, so EVERY field used for faceting/coloring must be grouped — else
        // VL drops it (an outer row facet would collapse to "undefined").
        const colorField = color?.field;
        const rowField = row?.field;
        const groupby = [catField];
        if (rowField && rowField !== catField) groupby.push(rowField);
        if (colorField && colorField !== catField && colorField !== rowField) groupby.push(colorField);
        spec.transform[0].density = measureField;
        spec.transform[0].groupby = groupby;

        // --- Bandwidth (mirrors density.ts) ---
        // Resolve the effective bandwidth FIRST so the extent padding below can
        // match the actual kernel width Vega will use.
        const config = ctx.chartProperties;
        const userBandwidth = config?.bandwidth && config.bandwidth > 0 ? config.bandwidth : 0;
        if (userBandwidth > 0) {
            spec.transform[0].bandwidth = userBandwidth;
        }

        // Evaluate every violin over the SAME measure range so they share the
        // value axis and stay directly comparable. The shared `extent` is what
        // Vega clips the KDE to, so it must reach far enough past the data that
        // the kernel tails decay to ~zero — otherwise the widest-spread group
        // (whose data fills the extent) ends in a flat clipped slab instead of a
        // tapered point. Pad by ~1.5 bandwidths past the data on each side
        // (seaborn-style "cut"), using the user bandwidth when set or the
        // normal-reference auto bandwidth Vega itself would pick.
        const extent = numericExtent(ctx.table, measureField);
        if (extent) {
            const range = extent[1] - extent[0];
            const effectiveBw = userBandwidth > 0
                ? userBandwidth
                : maxGroupBandwidth(ctx.table, measureField, groupby);
            const pad = Math.max(range * 0.05, 1.5 * effectiveBw, 1e-6);
            spec.transform[0].extent = [extent[0] - pad, extent[1] + pad];
        }

        // --- Value axis (the measure) ---
        spec.encoding.y.title = measureField;

        // --- Color: default to the category so each violin has its own hue ---
        const catType = isDiscrete(x?.type) ? x.type : 'nominal';
        if (colorField) {
            spec.encoding.color = { ...color };
        } else {
            spec.encoding.color = { field: catField, type: catType };
        }
        // The facet headers already label each category, so the color legend is
        // redundant when color mirrors the category.
        if (!colorField || colorField === catField) {
            spec.encoding.color.legend = null;
        }

        // --- Per-category panels: the category occupies the column/wrap facet ---
        const cats = distinctValues(ctx.table, catField);
        const catCount = Math.max(1, cats.length);

        // Size each violin panel from the canvas + category count (single row,
        // wrapping only when the row would get too cramped).
        const canvasW = ctx.canvasSize?.width ?? 560;
        const canvasH = ctx.canvasSize?.height ?? 360;
        const spacing = 0;
        const reservedW = 60;   // value axis + its title
        const reservedH = 70;   // facet headers + breathing room
        const minPanelW = 44;
        const maxPerRow = Math.max(1, Math.floor((canvasW - reservedW) / (minPanelW + spacing)));
        const columns = Math.min(catCount, maxPerRow);
        const rows = Math.ceil(catCount / columns);
        let panelW = Math.round((canvasW - reservedW - (columns - 1) * spacing) / columns);
        panelW = Math.max(minPanelW, Math.min(panelW, 220));
        const panelH = Math.max(120, Math.round((canvasH - reservedH) / rows) - (rows > 1 ? 24 : 0));

        const facetDef: any = {
            field: catField,
            type: catType,
            ...(x?.sort !== undefined ? { sort: x.sort } : {}),
            spacing,
            header: { titleOrient: 'bottom', labelOrient: 'bottom', labelPadding: 2 },
        };

        if (row) {
            // An additional OUTER facet → 2-D grid: category in columns, the
            // outer field in rows (VL cannot combine a wrap `facet` with `row`).
            // A non-wrap `column` + `row` pair is left intact by restructureFacets.
            spec.encoding.column = facetDef;
            spec.encoding.row = row;
        } else {
            // The per-category panels occupy a wrap facet with an EXPLICIT column
            // count. We set `encoding.facet` directly (not `encoding.column`) so
            // the assembler's restructureFacets leaves the grid untouched — it
            // would otherwise collapse an un-tracked column facet to `columns: 1`.
            spec.encoding.facet = { ...facetDef, columns };
        }

        // Explicit per-panel size (numbers) so vlApplyLayoutToSpec keeps them and
        // the grid is the per-category strip, not a single oversized plot.
        spec.width = panelW;
        spec.height = panelH;
    },
    properties: [
        { key: 'bandwidth', label: 'Bandwidth', type: 'continuous', min: 0.05, max: 2, step: 0.05, defaultValue: 0 },
    ] as ChartPropertyDef[],
};

File: packages/flint-py/flint/vegalite/templates/violin.py
"""Violin Plot template.

Shows the full smoothed distribution shape of a quantitative measure, one
mirrored kernel-density curve per category. Port of vegalite/templates/violin.ts.
"""
from __future__ import annotations

import math

from ...core import js_round
from .utils import detect_banded_axis_force_discrete


def _is_discrete(t):
    return t == "nominal" or t == "ordinal"


def _distinct_values(table, field):
    """Distinct non-null values of a field, in data-encounter order."""
    seen = set()
    out = []
    for row in table:
        v = row.get(field)
        if v is None:
            continue
        key = str(v)
        if key not in seen:
            seen.add(key)
            out.append(v)
    return out


def _numeric_extent(table, field):
    """[min, max] of a numeric field across the table (ignoring non-numbers)."""
    mn = math.inf
    mx = -math.inf
    for row in table:
        v = row.get(field)
        if not isinstance(v, (int, float)) or isinstance(v, bool) or not math.isfinite(v):
            continue
        if v < mn:
            mn = v
        if v > mx:
            mx = v
    if mn == math.inf or mx == -math.inf:
        return None
    if mn == mx:
        mn -= 0.5
        mx += 0.5
    return [mn, mx]


def _quantile_sorted(sorted_vals, p):
    """Linear-interpolated quantile of an already-sorted ascending array."""
    n = len(sorted_vals)
    if n == 0:
        return float("nan")
    idx = (n - 1) * p
    lo = math.floor(idx)
    hi = math.ceil(idx)
    if lo == hi:
        return sorted_vals[lo]
    return sorted_vals[lo] + (sorted_vals[hi] - sorted_vals[lo]) * (idx - lo)


def _stdev(values):
    """Sample standard deviation (n-1) of a numeric array."""
    n = len(values)
    if n < 2:
        return 0
    mean = sum(values) / n
    v = sum((x - mean) * (x - mean) for x in values) / (n - 1)
    return math.sqrt(v)


def _bandwidth_nrd(values):
    """Normal-reference (Scott/Silverman) KDE bandwidth."""
    n = len(values)
    if n < 2:
        return 0
    s = sorted(values)
    lo = _quantile_sorted(s, 0.25)
    hi = _quantile_sorted(s, 0.75)
    sd = _stdev(s)
    h = min(sd, (hi - lo) / 1.34)
    if not (h > 0):
        h = sd or abs(lo) or 1
    return 1.06 * h * (n ** -0.2)


def _max_group_bandwidth(table, measure, groupby):
    """Largest per-group KDE bandwidth across the groups Vega will form."""
    groups = {}
    for row in table:
        v = row.get(measure)
        if not isinstance(v, (int, float)) or isinstance(v, bool) or not math.isfinite(v):
            continue
        key = "\u0000".join(str(row.get(f)) for f in groupby)
        groups.setdefault(key, []).append(v)
    mx = 0
    for arr in groups.values():
        bw = _bandwidth_nrd(arr)
        if bw > mx:
            mx = bw
    return mx


def _violin_declare_layout(cs, table, chart_properties):
    if not (cs.get("x") or {}).get("field") or not (cs.get("y") or {}).get("field"):
        return {}
    result = detect_banded_axis_force_discrete(cs, table, {"preferAxis": "x"})
    if not result:
        return {}
    return {"resolvedTypes": result.get("resolvedTypes")}


def _violin_instantiate(spec, ctx):
    encs = ctx["resolvedEncodings"]
    x = encs.get("x")
    y = encs.get("y")
    color = encs.get("color")
    row = encs.get("row")
    cat_field = x.get("field") if x else None
    measure_field = y.get("field") if y else None
    if not cat_field or not measure_field:
        return

    # --- Density transform (per category, retaining every facet field) ---
    color_field = color.get("field") if color else None
    row_field = row.get("field") if row else None
    groupby = [cat_field]
    if row_field and row_field != cat_field:
        groupby.append(row_field)
    if color_field and color_field != cat_field and color_field != row_field:
        groupby.append(color_field)
    spec["transform"][0]["density"] = measure_field
    spec["transform"][0]["groupby"] = groupby

    # --- Bandwidth (mirrors density.ts) ---
    config = ctx.get("chartProperties") or {}
    user_bandwidth = config["bandwidth"] if config.get("bandwidth") and config["bandwidth"] > 0 else 0
    if user_bandwidth > 0:
        spec["transform"][0]["bandwidth"] = user_bandwidth

    extent = _numeric_extent(ctx["table"], measure_field)
    if extent:
        rng = extent[1] - extent[0]
        effective_bw = user_bandwidth if user_bandwidth > 0 else _max_group_bandwidth(ctx["table"], measure_field, groupby)
        pad = max(rng * 0.05, 1.5 * effective_bw, 1e-6)
        spec["transform"][0]["extent"] = [extent[0] - pad, extent[1] + pad]

    # --- Value axis (the measure) ---
    spec["encoding"]["y"]["title"] = measure_field

    # --- Color: default to the category so each violin has its own hue ---
    cat_type = x.get("type") if (x and _is_discrete(x.get("type"))) else "nominal"
    if color_field:
        spec["encoding"]["color"] = {**color}
    else:
        spec["encoding"]["color"] = {"field": cat_field, "type": cat_type}
    if not color_field or color_field == cat_field:
        spec["encoding"]["color"]["legend"] = None

    # --- Per-category panels: the category occupies the column/wrap facet ---
    cats = _distinct_values(ctx["table"], cat_field)
    cat_count = max(1, len(cats))

    canvas_w = (ctx.get("canvasSize") or {}).get("width") or 560
    canvas_h = (ctx.get("canvasSize") or {}).get("height") or 360
    spacing = 0
    reserved_w = 60
    reserved_h = 70
    min_panel_w = 44
    max_per_row = max(1, math.floor((canvas_w - reserved_w) / (min_panel_w + spacing)))
    columns = min(cat_count, max_per_row)
    rows_count = math.ceil(cat_count / columns)
    panel_w = js_round((canvas_w - reserved_w - (columns - 1) * spacing) / columns)
    panel_w = max(min_panel_w, min(panel_w, 220))
    panel_h = max(120, js_round((canvas_h - reserved_h) / rows_count) - (24 if rows_count > 1 else 0))

    facet_def = {
        "field": cat_field,
        "type": cat_type,
        "spacing": spacing,
        "header": {"titleOrient": "bottom", "labelOrient": "bottom", "labelPadding": 2},
    }
    if x and "sort" in x:
        facet_def["sort"] = x["sort"]

    if row:
        spec["encoding"]["column"] = facet_def
        spec["encoding"]["row"] = row
    else:
        spec["encoding"]["facet"] = {**facet_def, "columns": columns}

    spec["width"] = panel_w
    spec["height"] = panel_h


violin_plot_def = {
    "chart": "Violin Plot",
    "template": {
        "mark": {"type": "area", "orient": "horizontal"},
        "transform": [{"density": "__measure__", "groupby": [], "as": ["value", "density"]}],
        "encoding": {
            "y": {"field": "value", "type": "quantitative"},
            "x": {
                "field": "density", "type": "quantitative", "stack": "center",
                "impute": None, "title": None,
                "axis": {"labels": False, "ticks": False, "grid": False},
            },
        },
    },
    "channels": ["x", "y", "color", "row"],
    "markCognitiveChannel": "area",
    "declareLayoutMode": _violin_declare_layout,
    "instantiate": _violin_instantiate,
    "properties": [
        {"key": "bandwidth", "label": "Bandwidth", "type": "continuous",
         "min": 0.05, "max": 2, "step": 0.05, "defaultValue": 0},
    ],
}
